### DISCUSSION TITLE
NanoPi R6C Boot from NVMe - Idiots guide for a newbie to Linux...

### DISCUSSION THREAD

**Discussion by RicStrang** [Joshua-Riek/ubuntu-rockchip#257] [Q&A] [answered]
I followed the guide to boot my R6C Ubuntu Desktop from EMMC which works great, but would like to boot from NVMe or boot from EMMC but run the OS from the NVMe.

Any help would be much appreciated.

Many thanks

Richard

**Comment by Joshua-Riek** [comment]
I have a guide below on how to install the OS onto your eMMC:
https://github.com/Joshua-Riek/ubuntu-rockchip/wiki/FriendlyELEC-NanoPi-R6C#install-ubuntu-onto-your-emmc-from-linux

You can also run your OS on the NVMe by writing the bootloader to the eMMC. However, if you format the eMMC it will remove the bootloader.
```
sudo dd if=/lib/u-boot-nanopi-rk3588/idbloader.img of=/dev/mmcblk2 seek=64 conv=notrunc
sudo dd if=/lib/u-boot-nanopi-rk3588/u-boot.itb of=/dev/mmcblk2 seek=16384 conv=notrunc
```

**Reply by RicStrang** [reply]
Ok, thanks.

lsblk gives me this:

NAME         MAJ:MIN RM   SIZE RO TYPE MOUNTPOINTS
loop0          7:0    0  46.4M  1 loop /snap/snapd/19459
loop1          7:1    0  59.2M  1 loop /snap/core20/1977
loop2          7:2    0     4K  1 loop /snap/bare/5
loop3          7:3    0  91.7M  1 loop /snap/gtk-common-themes/1535
loop4          7:4    0 226.3M  1 loop /snap/firefox/2910
loop5          7:5    0   334M  1 loop /snap/gnome-3-38-2004/145
mmcblk2      179:0    0  28.9G  0 disk
├─mmcblk2p1  179:1    0   512M  0 part /boot/firmware
└─mmcblk2p2  179:2    0  28.4G  0 part /var/snap/firefox/common/host-hunspell
                                       /
mmcblk2boot0 179:32   0     4M  1 disk
mmcblk2boot1 179:64   0     4M  1 disk
nvme0n1      259:0    0 238.5G  0 disk
└─nvme0n1p1  259:1    0 238.5G  0 part

So, instead of - xz -dc ubuntu-22.04.2-preinstalled-desktop-arm64-nanopir6c.img.xz | sudo dd of=/dev/mmcblk2 bs=4k

Do I write to the NVMe as - xz -dc ubuntu-22.04.2-preinstalled-desktop-arm64-nanopir6c.img.xz | sudo dd of=/dev/nvme0n1 bs=4k

Then 
sudo dd if=/usr/lib/u-boot-nanopi-rk3588/idbloader.img of=/dev/mmcblk2 seek=64 conv=notrunc
sudo dd if=/usr/lib/u-boot-nanopi-rk3588/u-boot.itb of=/dev/mmcblk2 seek=16384 conv=notrunc

Files idbloader.img & u-boot.itb seem to be in /usr/lib/u-boot-nanopi-rk3588 for me...

**Reply by Joshua-Riek** [reply, marked as answer]
Correct. However, there is an issue with clashing UUID, so you must erase your eMMC first. 

I would boot from an SD card, then erase the eMMC and install it onto the NVMe. **Warning, the first dd command below will erase the eMMC.**
```
dd if=/dev/zero of=/dev/mmcblk2 count=4096 bs=512
xz -dc ubuntu-22.04.2-preinstalled-desktop-arm64-nanopir6c.img.xz | sudo dd of=/dev/nvme0n1 bs=4k
sudo dd if=/lib/u-boot-nanopi-rk3588/idbloader.img of=/dev/mmcblk2 seek=64 conv=notrunc
sudo dd if=/lib/u-boot-nanopi-rk3588/u-boot.itb of=/dev/mmcblk2 seek=16384 conv=notrunc
sync
```

**Reply by RicStrang** [reply]
That's great, works for me :-)

Many thanks

Any thoughts about booting direct from NVMe?

**Reply by Joshua-Riek** [reply]
The system will look for the bootloader on the SD Card then eMMC. It's how the system was designed, there is no way to **directly** boot from NVMe.

**Reply by amelchio** [reply]
I have the R6C without eMMC so I wanted to follow this setup with an SD card but it didn't boot.

Creating an empty partition table by running `sudo gpart /dev/sda` after the first step made it work for me. I used an external card reader so it was `/dev/sda` while preparing and is now `/dev/mmcblk1` when booted from the internal slot.

Also, the `/lib` paths were slightly different, like `/lib/u-boot/idbloader.img`.

Hope this helps someone, and thank you for helping me.

**Reply by dinjojo** [reply]
So whether you like it or not you need to have microsd card right ? I wanted to avoid using sd card as they get hot and eventually die so i was thinking of clearing emmc and then boot into nvme is it possible

**Reply by dinjojo** [reply]
> designed

Cna I skip sd card and boot directly from nvme

**Reply by harleycalvert** [reply]
You're going to need to clone your SD Card to your NVMe. The eMMC just adds an extra step. I ran my whole thing off an SD Card for about a year before I cloned it to NVMe. Now during the initial boot, it starts from SD then handballs everything to the NVMe. No problems so far. I'd skip the whole eMMC part. It just complicates things for no advantage.

**Reply by dinjojo** [reply]
I don't the lib directory in openwrt how do I clone it?

**Reply by harleycalvert** [reply]
what?

**Reply by dinjojo** [reply]
> what?

I'm on openwrt not on Debian is this only possible in Debian? /lib dir does not exists in openwrt

**Reply by harleycalvert** [reply]
I posted a how to here: https://www.friendlyelec.com/Forum/viewtopic.php?f=79&t=4241

It works for me.

**Reply by dinjojo** [reply]
It's for friendly wrt does openwrt also has same parition table also that's r5c

**Reply by harleycalvert** [reply]
It should be pretty much the same procedure. The partition numbers might differ. You just got to get in there and have good crack at it. Why not use the FriendlyWRT with Docker?

**Comment by harleycalvert** [comment]
I posted a how to here: https://www.friendlyelec.com/Forum/viewtopic.php?f=79&t=4241
Technically it boots from SD Card, but then the boot loader tells the NanoPi the root file system is on the NVMe and off it goes.
